crates/services/sync/src/import.rs

```diff
@@ -250,6 +229,54 @@ where
         .in_current_span()
         .await
     }
+
+    async fn get_block_for_header(
+        header: SourcePeer<SealedBlockHeader>,
+        p2p: Arc<P>,
+        consensus_port: Arc<C>,
+    ) -> anyhow::Result<Option<SealedBlock>> {
+        let SourcePeer {
+            peer_id,
+            data: header,
+        } = header;
+        let id = header.entity.id();
+        let block_id = SourcePeer { peer_id, data: id };
+
+        // Check the consensus is valid on this header.
+        if !consensus_port
+            .check_sealed_header(&header)
+            .trace_err("Failed to check consensus on header")?
+        {
+            tracing::warn!("Header {:?} failed consensus check", header);
+            return Ok(None)
+        }
+
+        // Wait for the da to be at least the da height on the header.
+        consensus_port
+            .await_da_height(&header.entity.da_height)
+            .await?;
+
+        get_transactions_on_block(p2p.as_ref(), block_id, header).await
+    }
+}
+
+// The response from the P2P network isn't guaranteed to include all the
+fn check_headers_match_range(
+    headers: &[SourcePeer<SealedBlockHeader>],
+    range: RangeInclusive<u32>,
+) -> Vec<SourcePeer<SealedBlockHeader>> {
+    let mut new_headers = Vec::new();
+    let mut header_iter = headers.iter();
+    for height in range {
+        if let Some(header) =
```

Comment:
It is a bad idea to use `find` because it is `O(n^2)` in the worst case.

You know that each block should have `i == header.height`, so you can iterate over the range and compare the height. You finish the iteration if you don't have more elements or one has an unexpected height. All valid block headers you can use for future processing. For all remaining blocks, you need to send a request again, and report about invalid behavior of the peer(we don't do that right now, but you can leave a TODO).


The right ordering(from smallest to biggest) of the block header should be a rule that each node should follow.

Reply:
I agree on ordering. There is no danger to them returning the wrong order, it just will error on the check.